cumulusci/tests/util.py

```diff
@@ -87,3 +95,152 @@ def global_config_dir(self):
     @property
     def cache_dir(self):
         return Path.home() / "project/.cci"
+
+
+@contextmanager
+def assert_max_memory_usage(max_usage):
+    "Assert that a test does not exceed a certain memory threshold"
+    tracemalloc.start()
+    yield
+    current, peak = tracemalloc.get_traced_memory()
+    if peak > max_usage:
+        big_objs(traced_only=True)
+        assert peak < max_usage, ("Peak incremental memory usage was high:", peak)
+    tracemalloc.stop()
+
+
+def big_objs(traced_only=False):
+
+    big_objs = (
+        (sys.getsizeof(obj), obj)
+        for obj in gc.get_objects()
+        if sys.getsizeof(obj) > 20000
+        and (tracemalloc.get_object_traceback(obj) if traced_only else True)
+    )
+    for size, obj in big_objs:
+        print(type(obj), size, tracemalloc.get_object_traceback(obj))
+
+
+class MockBulkAPI:
+    """Extremely simplistic mock of the bulk API
+
+    Can be improved as needed over time.
+    """
+
+    next_job_id = 0
+    next_batch_id = 0
+
+    @classmethod
+    def create_job(cls, *args, **kwargs):
+        cls.next_job_id += 1
+        return cls.next_job_id
+
+    @classmethod
+    def create_query_job(cls, *args, **kwargs):
+        cls.next_job_id += 1
+        return cls.next_job_id
+
+    @classmethod
+    def post_batch(cls, *args, **kwargs):
+        cls.next_batch_id += 1
+        return cls.next_batch_id
+
+    def close_job(self, *args, **kwargs):
+        pass
+
+    def job_status(self, job_id):
+        return {
+            "numberBatchesCompleted": self.next_batch_id,
+            "numberBatchesTotal": self.next_batch_id,
+        }
+
+    def query(self, job_id, soql):
+        return self.post_batch()
+
+    def get_query_batch_result_ids(self, batch_id, job_id):
+        return range(0, 10)
+
+
+class MockSF:
+    """Extremely simplistic mock of the Salesforce API
+
+    Can be improved as needed over time.
+    """
+
+    mocks = {}
+
+    def describe(self):
+        return getattr(self, "global_describe")
+
+    @property
+    def sf_version(self):
+        return 47
+
+    def __getattr__(self, sobjname):
+        self.mocks[sobjname] = self.mocks.get("sobjname", None) or read_mock(sobjname)
+        return self.mocks[sobjname]
```

Comment:
Err...this is just returning a json string, right? Which is not what the real simple-salesforce `__getattr__` does.

Reply:
Renamed.